Adding SegmentMetadataEvent and publishing them via KafkaEmitter

harinirajendran opened the pull request:
## Details
Adding the new SegmentMetadataEvent and publishing these segment-related metadata events into Kafka by enhancing the KafkaEmitter

## Description
In this PR, we are enhancing KafkaEmitter, to emit metadata about published segments (SegmentMetadataEvent) into a Kafka topic. This segment metadata information that gets published into Kafka, can be used by any other downstream services to query Druid intelligently based on the segments published. The segment metadata gets published into kafka topic in json string format similar to other events.

## Old behavior of Kafka Emitter
Kafka Emitter always emits metrics and alerts and would emit requests if the config request.topic is configured.
Configs metric.topic and alert.topic are always mandatory and cannot be null.

## Current behavior of Kafka Emitter [with backwards compatibility]
We introduced a new config named event.types which dictates the types of events we want the KafkaEmitter to emit. This config takes in a list of strings and can have one or more from [alerts, metrics, requests and segmentMetadata]. And based on this config, alert.topic, request.topic, metric.topic and segmentMetadata.topic should be configured and not left empty.
If no event.types is set, then by default, the kafka emitter would emit metrics and alerts. And in that case, to maintain backwards compatibility, decision to send out requests are based on if request.topic is empty or set.

This PR has:

- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

harinirajendran commented:
@nishantmonu51 :  I have added a new event type in this PR that would cause the following exception to be thrown in AmbariMetricsEmitter and DropWizardEmitter
```
throw new ISE("unknown event type [%s]", event.getClass());
```
Should I explicitly filter this new event type from these emitters?

gianm commented:
> Should I explicitly filter this new event type from these emitters?

IMO it would make more sense to edit those emitters to ignore unknown event types.

Btw, this may be OK for your use case, but, I wanted to point out that there is a race here: it's possible for the segments to be committed and for the server to crash before it emits this event. So, some might get missed.

harinirajendran commented:
> IMO it would make more sense to edit those emitters to ignore unknown event types.

That makes sense @gianm. I'll update those emitters to ignore the new event type.

harinirajendran commented:
> Btw, this may be OK for your use case, but, I wanted to point out that there is a race here: it's possible for the segments to be committed and for the server to crash before it emits this event. So, some might get missed.

That's a great point @gianm. Is there a better place to emit this segment metadata event instead of this place to prevent this?

abhishekrb19 commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.alert.topic`                  | Kafka topic name for emitter's target to emit alert. If `event.types` contains `alerts`, this field cannot be left empty                                                                | no        | none                  |
+| `druid.emitter.kafka.request.topic`                | Kafka topic name for emitter's target to emit request logs. If `event.types` contains `requests`, this field cannot be left empty                                                       | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segments related metadata. If `event.types` contains `segmentMetadata`, this field cannot be left empty                                   | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic.format` | Format in which segment related metadata will be emitted. <br/>Choices: json, protobuf.<br/> If set to `protobuf`, then segment metadata is emitted in `DruidSegmentEvent.proto` format | no        | json                  |
```
I'm curious about the need to add protobuf encoding (and thereby a dependency) here when we can get away with `json` or a byte format? The output format, including json, can always be made backwards compatible.  And if downstream consumers want to consume the topics as a proto encoded message, that should still be possible by unmarshaling json/bytes into a proto struct as needed?

harinirajendran replied:
For now, we have decided to just stick with the JSON implementation like other events. I'll update the PR.

gianm commented:
> That's a great point @gianm. Do you have a recommendation for a better place to emit this segment metadata event instead of this place to prevent this?

I think for it to be "perfect" the best way to do it would be to emit in the place you emit here, but also have some other process that detects missed emits somehow and fixes them up by redoing the missed emits. This would be a lot more complex of an implementation, however. So I'd only recommend doing that if it seems worth it.

To figure that out, I would consider the requirements here. What kind of things are likely to consume the emitted payloads? Could they tolerate either of the following conditions?

- missed emits (segments that are published, but never emitted)
- bogus emits (segments that are never published, but were emitted anyway)

If one or both of these can be tolerated, the implementation becomes a lot simpler.

If I understand correctly— the one you have in this PR is the "missed emit" scenario. It won't generate bogus emits, but it can potentially miss some.

harinirajendran commented:
> I think for it to be "perfect" the best way to do it would be to emit in the place you emit here, but also have some other process that detects missed emits somehow and fixes them up by redoing the missed emits. This would be a lot more complex of an implementation, however. So I'd only recommend doing that if it seems worth it.

I think for now we can just go with this implement and enhance it to emit missed segments metadata in the future.

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
````diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+druid.emitter.kafka.segmentMetadata.topic.format=protobuf 
 druid.emitter.kafka.producer.config={"max.block.ms":10000}
 ```
+Whenever `druid.emitter.kafka.segmentMetadata.topic.format` field is updated, it is recommended to also update  `druid.emitter.kafka.segmentMetadata.topic` to avoid the same topic from getting polluted with different formats of segment metadata.
````
```suggestion
When you update `druid.emitter.kafka.segmentMetadata.topic.format`, it is recommended that you also update  `druid.emitter.kafka.segmentMetadata.topic` to avoid the same topic from getting polluted with different formats of segment metadata.
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
-|`druid.emitter.kafka.request.topic`|Kafka topic name for emitter's target to emit request logs. If left empty then request logs will not be sent to the Kafka topic.|no|none|
-|`druid.emitter.kafka.producer.config`|JSON formatted configuration which user want to set additional properties to Kafka producer.|no|none|
-|`druid.emitter.kafka.clusterName`|Optional value to specify name of your druid cluster. It can help make groups in your monitoring environment. |no|none|
+| property                                           | description                                                                                                                                                                             | required? | default               |
```
```suggestion
| Property                                           | Description                                                                                                                                                                             | Required | Default               |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
-|`druid.emitter.kafka.clusterName`|Optional value to specify name of your druid cluster. It can help make groups in your monitoring environment. |no|none|
+| property                                           | description                                                                                                                                                                             | required? | default               |
+|----------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|-----------------------|
+| `druid.emitter.kafka.bootstrap.servers`            | Comma-separated Kafka broker. (`[hostname:port],[hostname:port]...`)                                                                                                                    | yes       | none                  |
```
```suggestion
| `druid.emitter.kafka.bootstrap.servers`            | Comma-separated Kafka broker. (`[hostname:port],[hostname:port]...`)                                                                                                                    | Yes       | None                  |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+|----------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|-----------------------|
+| `druid.emitter.kafka.bootstrap.servers`            | Comma-separated Kafka broker. (`[hostname:port],[hostname:port]...`)                                                                                                                    | yes       | none                  |
+| `druid.emitter.kafka.event.types`                  | Comma-separated event types. <br/>Choices: alerts, metrics, requests, segmentMetadata                                                                                                   | no        | ["metrics", "alerts"] |
+| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metric. If `event.types` contains `metrics`, this field cannot be left empty                                                      | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metric. If `event.types` contains `metrics`, this field cannot be empty.                                                     | no        | none                  |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.bootstrap.servers`            | Comma-separated Kafka broker. (`[hostname:port],[hostname:port]...`)                                                                                                                    | yes       | none                  |
+| `druid.emitter.kafka.event.types`                  | Comma-separated event types. <br/>Choices: alerts, metrics, requests, segmentMetadata                                                                                                   | no        | ["metrics", "alerts"] |
+| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metric. If `event.types` contains `metrics`, this field cannot be left empty                                                      | no        | none                  |
+| `druid.emitter.kafka.alert.topic`                  | Kafka topic name for emitter's target to emit alert. If `event.types` contains `alerts`, this field cannot be left empty                                                                | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.alert.topic`                  | Kafka topic name for emitter's target to emit alerts. If `event.types` contains `alerts`, this field cannot empty.                                                                | no        | none                  |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.event.types`                  | Comma-separated event types. <br/>Choices: alerts, metrics, requests, segmentMetadata                                                                                                   | no        | ["metrics", "alerts"] |
+| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metric. If `event.types` contains `metrics`, this field cannot be left empty                                                      | no        | none                  |
+| `druid.emitter.kafka.alert.topic`                  | Kafka topic name for emitter's target to emit alert. If `event.types` contains `alerts`, this field cannot be left empty                                                                | no        | none                  |
+| `druid.emitter.kafka.request.topic`                | Kafka topic name for emitter's target to emit request logs. If `event.types` contains `requests`, this field cannot be left empty                                                       | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.request.topic`                | Kafka topic name for emitter's target to emit request logs. If `event.types` contains `requests`, this field cannot be empty.                                                       | no        | none                  |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metric. If `event.types` contains `metrics`, this field cannot be left empty                                                      | no        | none                  |
+| `druid.emitter.kafka.alert.topic`                  | Kafka topic name for emitter's target to emit alert. If `event.types` contains `alerts`, this field cannot be left empty                                                                | no        | none                  |
+| `druid.emitter.kafka.request.topic`                | Kafka topic name for emitter's target to emit request logs. If `event.types` contains `requests`, this field cannot be left empty                                                       | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segments related metadata. If `event.types` contains `segmentMetadata`, this field cannot be left empty                                   | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segments related metadata. If `event.types` contains `segmentMetadata`, this field cannot be empty.                                   | no        | none                  |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.request.topic`                | Kafka topic name for emitter's target to emit request logs. If `event.types` contains `requests`, this field cannot be left empty                                                       | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segments related metadata. If `event.types` contains `segmentMetadata`, this field cannot be left empty                                   | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic.format` | Format in which segment related metadata will be emitted. <br/>Choices: json, protobuf.<br/> If set to `protobuf`, then segment metadata is emitted in `DruidSegmentEvent.proto` format | no        | json                  |
+| `druid.emitter.kafka.producer.config`              | JSON formatted configuration which user want to set additional properties to Kafka producer.                                                                                            | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.producer.config`              | JSON formatted configuration to set additional properties to Kafka producer.                                                                                            | no        | none                  |
```

ektravel reviewed (commented):
Docs: left some questions/suggestions.

harinirajendran commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,28 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segments related metadata. If `event.types` contains `segmentMetadata`, this field cannot be left empty                                   | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic.format` | Format in which segment related metadata will be emitted. <br/>Choices: json, protobuf.<br/> If set to `protobuf`, then segment metadata is emitted in `DruidSegmentEvent.proto` format | no        | json                  |
+| `druid.emitter.kafka.producer.config`              | JSON formatted configuration which user want to set additional properties to Kafka producer.                                                                                            | no        | none                  |
+| `druid.emitter.kafka.clusterName`                  | Optional value to specify name of your druid cluster. It can help make groups in your monitoring environment.                                                                           | no        | none                  |
```
We can group the metrics by clustername to differentiate them from one cluster to another in monitoring environments like datadog, new relic, etc.

abhishekrb19 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java:
```diff
@@ -21,53 +21,115 @@
     this.kafkaProducerConfig = kafkaProducerConfig == null ? ImmutableMap.of() : kafkaProducerConfig;
   }
 
+  private Set<EventType> validateEventTypes(Set<EventType> eventTypes, String requestTopic)
```
Should we rename this function since it's not really validating the supplied event types? Or maybe have a default jackson getter for the event types property that bakes in the backwards compatibility logic?

abhishekrb19 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java:
```diff
@@ -105,7 +111,7 @@ protected Producer<String, String> setKafkaProducer()
       props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, config.getBootstrapServers());
       props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
-      props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
+      props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, ByteArraySerializer.class.getName());
```
I think this would be a breaking change for existing consumers that expect string types.

abhishekrb19 commented on extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:
```diff
@@ -77,19 +84,31 @@ public void testKafkaEmitter() throws InterruptedException
+    final List<SegmentMetadataEvent> segmentMetadataEvents = ImmutableList.of(
+        new SegmentMetadataEvent(
+            "dummy_datasource",
+            DateTimes.of("2001-01-01T00:00:00.000Z"),
```
👍

harinirajendran commented on .idea/misc.xml:
```diff
@@ -40,6 +40,11 @@
     </option>
+    <option name="ignoredFiles">
+      <set>
+        <option value="$PROJECT_DIR$/integration-tests/pom.xml" />
```
Removing them.

harinirajendran reviewed (commented):
Addressed all the review comments

abhishekrb19 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java:
```diff
@@ -183,24 +202,31 @@ public void emit(final Event event)
           invalidLost.incrementAndGet();
         }
       }
-      catch (JsonProcessingException e) {
```
Do we need this wider catch block or should `JsonProcessingException` suffice?

harinirajendran replied:
Changed it back

abhishekrb19 commented on extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:
```diff
@@ -47,20 +51,23 @@
     };
   }
 
   // there is 1 seconds wait in kafka emitter before it starts sending events to broker, set a timeout for 5 seconds
```
Do we see this test fail sometimes without this bump in timeout?

```suggestion
  // there is 1 seconds wait in kafka emitter before it starts sending events to broker, set a timeout for 10 seconds
```

harinirajendran replied:
yes. atleast in my laptop it did fail a few times.

abhishekrb19 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java:
```diff
@@ -21,53 +21,114 @@
+    METRICS,
+    ALERTS,
+    REQUESTS,
+    SEGMENTMETADATA {
```
nit: for readability, I'd suggest `SEGMENT_METADATA`. Also, I think with that we won't need a custom enum `toString()`

abhishekrb19 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java:
```diff
@@ -183,24 +202,31 @@ public void emit(final Event event)
-      catch (JsonProcessingException e) {
+      catch (Exception e) {
         invalidLost.incrementAndGet();
+        log.warn(e, "Exception while serializing event");
```
Related to the above comment, this log line only applies to `JsonProcessingException`s. I wonder if we should log it as an error and make it general enough instead of "serializing event".

harinirajendran replied:
now that we are not doing protobuf, I changed it back to JsonProcessingException. So, leaving this line as is.

abhishekrb19 reviewed (commented):
Thank you, @harinirajendran! Looks good - I only have a few minor and optional comments

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,26 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metric. If `event.types` contains `metrics`, this field cannot be empty.                                                     | no        | none                  |
+| `druid.emitter.kafka.alert.topic`                  | Kafka topic name for emitter's target to emit alerts. If `event.types` contains `alerts`, this field cannot empty.                                                                | no        | none                  |
+| `druid.emitter.kafka.request.topic`                | Kafka topic name for emitter's target to emit request logs. If `event.types` contains `requests`, this field cannot be empty.                                                       | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segments related metadata. If `event.types` contains `segmentMetadata`, this field cannot be empty.                                   | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segment metadata. If `event.types` contains `segmentMetadata`, this field cannot be empty.                                   | no        | none                  |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,26 @@ to monitor the status of your Druid cluster with this extension.
+|----------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------|-----------------------|
+| `druid.emitter.kafka.bootstrap.servers`            | Comma-separated Kafka broker. (`[hostname:port],[hostname:port]...`)                                                                                                                    | yes       | none                  |
+| `druid.emitter.kafka.event.types`                  | Comma-separated event types. <br/>Supported types are `alerts`, `metrics`, `requests`, and `segmentMetadata`.                                                                                                   | no        | `["metrics", "alerts"]` |
+| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metric. If `event.types` contains `metrics`, this field cannot be empty.                                                     | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.metric.topic`                 | Kafka topic name for emitter's target to emit service metrics. If `event.types` contains `metrics`, this field cannot be empty.                                                     | no        | none                  |
```

ektravel commented on docs/development/extensions-contrib/kafka-emitter.md:
```diff
@@ -36,20 +36,26 @@ to monitor the status of your Druid cluster with this extension.
+| `druid.emitter.kafka.alert.topic`                  | Kafka topic name for emitter's target to emit alerts. If `event.types` contains `alerts`, this field cannot empty.                                                                | no        | none                  |
+| `druid.emitter.kafka.request.topic`                | Kafka topic name for emitter's target to emit request logs. If `event.types` contains `requests`, this field cannot be empty.                                                       | no        | none                  |
+| `druid.emitter.kafka.segmentMetadata.topic`        | Kafka topic name for emitter's target to emit segments related metadata. If `event.types` contains `segmentMetadata`, this field cannot be empty.                                   | no        | none                  |
+| `druid.emitter.kafka.producer.config`              | JSON formatted configuration to set additional properties to Kafka producer.                                                                                            | no        | none                  |
```
```suggestion
| `druid.emitter.kafka.producer.config`              | JSON configuration to set additional properties to Kafka producer.                                                                                            | no        | none                  |
```

ektravel reviewed (commented):
[Docs] A few nits but looks good otherwise.

ektravel reviewed (approved):
Looks good from the docs standpoint.

abhishekrb19 reviewed (approved):
LGTM (non-binding). Thanks for the neat patch, @harinirajendran!

abhishekrb19 commented:
I see UTs fail because of the enum change `SEGMENT_METADATA`. It appears that we only do upper and lower case conversions in `KafkaEmitterConfig.EventType#fromString` and `#toString` methods, so I think a custom `toString()` on the enum is needed?
```
Error:    Run 4: KafkaEmitterConfigTest.testDeserializeEventTypesWithDifferentCase:108 » ValueInstantiation Cannot construct instance of `org.apache.druid.emitter.kafka.KafkaEmitterConfig$EventType`, problem: No enum constant org.apache.druid.emitter.kafka.KafkaEmitterConfig.EventType.SEGMENTMETADATA
 at [Source: (String)""segmentMetadata""; line: 1, column: 1]
```

harinirajendran commented:
> I see UTs fail because of the enum change `SEGMENT_METADATA`. It appears that we only do upper and lower case conversions in `KafkaEmitterConfig.EventType#fromString` and `#toString` methods, so I think a custom `toString()` on the enum is needed?
> 
> ```
> Error:    Run 4: KafkaEmitterConfigTest.testDeserializeEventTypesWithDifferentCase:108 » ValueInstantiation Cannot construct instance of `org.apache.druid.emitter.kafka.KafkaEmitterConfig$EventType`, problem: No enum constant org.apache.druid.emitter.kafka.KafkaEmitterConfig.EventType.SEGMENTMETADATA
>  at [Source: (String)""segmentMetadata""; line: 1, column: 1]
> ```

I updated the event name from SEGMENTMETADATA to SEGMENT_METADATA but I missed to update the unit test I added. Hence the failure. Fixed it now.

harinirajendran commented:
@abhishekrb19 can u approve the PR again so that the build would get triggered?

abhishekrb19 commented:
@harinirajendran, CI gets triggered by committers - I've posted this in the dev slack channel, so someone should hopefully get to it soon.

harinirajendran commented:
@abhishekrb19 Can you give a +1 again? Fixed a bunch of static check failures.
<img width="1344" alt="Screen Shot 2023-05-26 at 8 51 51 AM" src="https://github.com/apache/druid/assets/9054348/1067d4c7-c85d-4346-be96-8a819c02f6f5">

xvrl commented:
@harinirajendran can we update the PR description to reflect the current state of the code. It still references the protobuf implementation which is no longer part of this.

xvrl commented on extensions-contrib/kafka-emitter/pom.xml:
```diff
@@ -86,7 +86,6 @@
       <artifactId>slf4j-api</artifactId>
       <scope>provided</scope>
     </dependency>
-
```
nit, can we avoid unnecessary formatting changes here since we're not making change to this file?

xvrl commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java:
```diff
@@ -173,8 +193,8 @@ public void emit(final Event event)
           map = map.asBuilder()
-                   .put("clusterName", config.getClusterName())
-                   .build();
+              .put("clusterName", config.getClusterName())
+              .build();
```
nit, the formatting change here seems unnecessary

xvrl commented on indexing-service/src/main/java/org/apache/druid/indexing/common/actions/SegmentTransactionalInsertAction.java:
```diff
@@ -257,11 +260,27 @@ public SegmentPublishResult perform(Task task, TaskActionToolbox toolbox)
           segment.getShardSpec() == null ? null : segment.getShardSpec().getType()
       );
       toolbox.getEmitter().emit(metricBuilder.build("segment/added/bytes", segment.getSize()));
+      // Emit the segment related metadata using the configured emitters
+      this.emitSegmentMetadata(segment, toolbox);
```
can we add a comment that captures the limitation that @gianm pointed out, explaining that some segmentmetadata events may not be captured in the event of server crash?

xvrl reviewed (commented):
a few minor nits, otherwise LGTM. Can we update the PR description to remove the mention of protobuf and also highlight the limitations in terms of delivery guarantees for those segments in the event that the server crashes?

abhishekagarwal87 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java:
```diff
@@ -119,18 +126,26 @@ protected Producer<String, String> setKafkaProducer()
     scheduler.scheduleWithFixedDelay(() -> {
-      log.info(
-          "Message lost counter: metricLost=[%d], alertLost=[%d], requestLost=[%d], invalidLost=[%d]",
+      log.info("Message lost counter: metricLost=[%d], alertLost=[%d], requestLost=[%d], invalidLost=[%d] segmentMetadataLost=[%d]",
```
nit. lets have invalidLost at the end.

abhishekagarwal87 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitter.java:
```diff
@@ -183,24 +203,31 @@ public void emit(final Event event)
+        Set<EventType> eventTypes = config.getEventTypes();
         if (event instanceof ServiceMetricEvent) {
-          if (!metricQueue.offer(objectContainer)) {
+          if (!eventTypes.contains(EventType.METRICS) || !metricQueue.offer(objectContainer)) {
```
why do we want to increment metricsLost in this case? Since not reporting is intentional.

abhishekagarwal87 replied:
Got it. gotta live with the sins of the past.

abhishekagarwal87 replied:
Got it. gotta live with the sins of the past.

abhishekagarwal87 commented on extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java:
```diff
@@ -21,53 +21,108 @@
-    this.alertTopic = Preconditions.checkNotNull(alertTopic, "alert.topic can not be null");
-    this.requestTopic = requestTopic;
+    this.eventTypes = maybeUpdateEventTypes(eventTypes, requestTopic);
+    this.metricTopic = this.eventTypes.contains(EventType.METRICS) ? Preconditions.checkNotNull(metricTopic, "metric.topic can not be null") : null;
```
unrelated to your PR but can the error message include the full config name. E.g. `druid.emitter.kafka.metric.topic can not be null`

abhishekagarwal87 commented on processing/src/main/java/org/apache/druid/java/util/emitter/service/SegmentMetadataEvent.java:
```diff
@@ -0,0 +1,81 @@
+import org.apache.druid.java.util.emitter.core.EventMap;
+import org.joda.time.DateTime;
+
+public class SegmentMetadataEvent implements Event
```
please add javadocs about the class.

harinirajendran commented:
> @harinirajendran can we update the PR description to reflect the current state of the code. It still references the protobuf implementation which is no longer part of this.

Updated it @xvrl . will address other comments now

harinirajendran reviewed (commented):
Addressed review comments.

abhishekagarwal87 commented:
@harinirajendran - We will have to fix it. When you run locally, are you running just one test? I will suggest running the same maven command that github action is running. I think that you will run into the same error then.

harinirajendran commented:
> @harinirajendran - We will have to fix it. When you run locally, are you running just one test? I will suggest running the same maven command that github action is running. I think that you will run into the same error then.

Yeah I was just running that one test. let me try using the same command and see if I can reproduce it. Thanks @abhishekagarwal87

abhishekagarwal87 commented:
@harinirajendran - You should avoid `force-push` in the future as we can't see the diff from the last commit anymore. can you describe the most recent change?

harinirajendran commented:
> @harinirajendran - You should avoid `force-push` in the future as we can't see the diff from the last commit anymore. can you describe the most recent change?

Ahh okay! Wasn't aware of that. Will keep that in mind for future PRs. I did not make any code changes. Just rebased the code on top of the latest master and pushed it again.
